pghoard/rohmu/delta/snapshot.py

```diff
@@ -0,0 +1,231 @@
+        return SnapshotState(
+            root_globs=self.globs, files=sorted(self.relative_path_to_snapshotfile.values()), empty_dirs=self.empty_dirs
+        )
+
+    def _snapshot_create_missing_directories(self, *, src_dirs, dst_dirs):
+        changes = 0
+        for i, relative_dir in enumerate(set(src_dirs).difference(dst_dirs), 1):
+            dst_path = self.dst / relative_dir
+            dst_path.mkdir(parents=True, exist_ok=True)
+            if increase_worth_reporting(i):
+                logger.debug("#%d. new directory: %r", i, relative_dir)
+            changes += 1
+        return changes
+
+    def _snapshot_remove_extra_files(self, *, src_files, dst_files):
+        changes = 0
+        for i, relative_path in enumerate(set(dst_files).difference(src_files), 1):
+            dst_path = self.dst / relative_path
+            snapshotfile = self.relative_path_to_snapshotfile.get(relative_path)
+            if snapshotfile:
+                self._remove_snapshotfile(snapshotfile)
+            dst_path.unlink()
+            if increase_worth_reporting(i):
+                logger.debug("#%d. extra file: %r", i, relative_path)
+            changes += 1
+        return changes
+
+    def _snapshot_add_missing_files(self, *, src_files, dst_files):
+        existing = 0
+        disappeared = 0
+        changes = 0
+        for i, relative_path in enumerate(set(src_files).difference(dst_files), 1):
+            src_path = self.src / relative_path
+            dst_path = self.dst / relative_path
+            try:
+                os.link(src=src_path, dst=dst_path, follow_symlinks=False)
+            except FileExistsError:
+                # This happens only if snapshot is started twice at
+                # same time. While it is technically speaking upstream
+                # error, we rather handle it here than leave
+                # exceptions not handled.
+                existing += 1
+                if increase_worth_reporting(existing):
+                    logger.debug("#%d. %s already existed, ignoring", existing, src_path)
+                continue
+            except FileNotFoundError:
+                disappeared += 1
+                if increase_worth_reporting(disappeared):
+                    logger.debug("#%d. %s disappeared before linking, ignoring", disappeared, src_path)
+                continue
+            if increase_worth_reporting(i - disappeared):
+                logger.debug("#%d. new file: %r", i - disappeared, relative_path)
+            changes += 1
+        return changes
+
+    def snapshot(self, *, progress: Optional[Progress] = None):
+        assert self.lock.locked()
+
+        if progress is None:
+            progress = Progress()
+        progress.start(3)
+
+        if self.src_iterate_func:
+            src_dirs = set()
+            src_files = set()
+            for item in self.src_iterate_func():
+                if os.path.isfile(item) and not os.path.islink(item):
```

Comment:
nit: what do you think about:
```
for item...:
    path = Path(item)
    if path.is_file...:
        src_files.add(item.relative_to(...
...
```
so, there will be uniform usage of `Path` along the code, as also code doesn't assume that `src_iterate_func()` generates strings (and less conversions/usage of misc. libs).